### PR TITLE
RETURNDATACOPY and RETURNDATASIZE

### PR DESCRIPTION
**Pull request by tinchou**
Implements `RETURNDATACOPY` and `RETURNDATASIZE` opcodes.

Based on EIP 211: https://github.com/ethereum/EIPs/blob/master/EIPS/eip-211.md
Based implementation on https://github.com/ethereum/ethereumj/pull/919

### CODE REVIEW

**Comment by rskops** [comment]
SonarQube analysis reported 410 issues
* ![MAJOR][MAJOR] 85 major
* ![MINOR][MINOR] 266 minor
* ![INFO][INFO] 59 info

#### Top 10 issues

1. ![MAJOR][MAJOR] [VM.java#](https://github.com/rsksmart/rskj/blob/887939af0aa74746c40188eef158e30a4f66b3ff/rskj-core/src/main/java/org/ethereum/vm/VM.java#): This file has 1,415 lines, which is greater than 1,000 authorized. Split it into smaller files. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS00104)
1. ![MAJOR][MAJOR] [VM.java#L106](https://github.com/rsksmart/rskj/blob/887939af0aa74746c40188eef158e30a4f66b3ff/rskj-core/src/main/java/org/ethereum/vm/VM.java#L106): Rename "gasCost" which hides the field declared at line 162. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AHiddenFieldCheck)
1. ![MAJOR][MAJOR] [VM.java#L180](https://github.com/rsksmart/rskj/blob/887939af0aa74746c40188eef158e30a4f66b3ff/rskj-core/src/main/java/org/ethereum/vm/VM.java#L180): Either log or rethrow this exception. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1166)
1. ![MAJOR][MAJOR] [VM.java#L206](https://github.com/rsksmart/rskj/blob/887939af0aa74746c40188eef158e30a4f66b3ff/rskj-core/src/main/java/org/ethereum/vm/VM.java#L206): Method org.ethereum.vm.VM.doADD() appears to call the same method on the same object redundantly [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3APRMC_POSSIBLY_REDUNDANT_METHOD_CALLS)
1. ![MAJOR][MAJOR] [VM.java#L222](https://github.com/rsksmart/rskj/blob/887939af0aa74746c40188eef158e30a4f66b3ff/rskj-core/src/main/java/org/ethereum/vm/VM.java#L222): Method org.ethereum.vm.VM.doMUL() appears to call the same method on the same object redundantly [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3APRMC_POSSIBLY_REDUNDANT_METHOD_CALLS)
1. ![MAJOR][MAJOR] [VM.java#L237](https://github.com/rsksmart/rskj/blob/887939af0aa74746c40188eef158e30a4f66b3ff/rskj-core/src/main/java/org/ethereum/vm/VM.java#L237): Method org.ethereum.vm.VM.doSUB() appears to call the same method on the same object redundantly [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3APRMC_POSSIBLY_REDUNDANT_METHOD_CALLS)
1. ![MAJOR][MAJOR] [VM.java#L252](https://github.com/rsksmart/rskj/blob/887939af0aa74746c40188eef158e30a4f66b3ff/rskj-core/src/main/java/org/ethereum/vm/VM.java#L252): Method org.ethereum.vm.VM.doDIV() appears to call the same method on the same object redundantly [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3APRMC_POSSIBLY_REDUNDANT_METHOD_CALLS)
1. ![MAJOR][MAJOR] [VM.java#L267](https://github.com/rsksmart/rskj/blob/887939af0aa74746c40188eef158e30a4f66b3ff/rskj-core/src/main/java/org/ethereum/vm/VM.java#L267): Method org.ethereum.vm.VM.doSDIV() appears to call the same method on the same object redundantly [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3APRMC_POSSIBLY_REDUNDANT_METHOD_CALLS)
1. ![MAJOR][MAJOR] [VM.java#L282](https://github.com/rsksmart/rskj/blob/887939af0aa74746c40188eef158e30a4f66b3ff/rskj-core/src/main/java/org/ethereum/vm/VM.java#L282): Method org.ethereum.vm.VM.doMOD() appears to call the same method on the same object redundantly [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3APRMC_POSSIBLY_REDUNDANT_METHOD_CALLS)
1. ![MAJOR][MAJOR] [VM.java#L297](https://github.com/rsksmart/rskj/blob/887939af0aa74746c40188eef158e30a4f66b3ff/rskj-core/src/main/java/org/ethereum/vm/VM.java#L297): Method org.ethereum.vm.VM.doSMOD() appears to call the same method on the same object redundantly [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3APRMC_POSSIBLY_REDUNDANT_METHOD_CALLS)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: rskj-core/src/main/java/org/ethereum/vm/VM.java [hunk 1]
```diff
@@ -699,16 +699,8 @@ protected void doCALLDATASIZE() {
...
-        long newMemSize ;
-        long copySize;
-
         if (computeGas) {
-            size = stack.get(stack.size() - 3);
-            copySize = Program.limitToMaxLong(size);
-            checkSizeArgument(copySize);
-            newMemSize = memNeeded(stack.peek(), copySize);
-            gasCost += calcMemGas(oldMemSize, newMemSize, copySize);
+            gasCost += computeDataCopyGas();
```
Comment (ajlopezrsk) [inline rskj-core/src/main/java/org/ethereum/vm/VM.java]: Any test to check the new gas cost for this opcode and the new RETURNDATACOPY opcode?

File: rskj-core/src/main/java/org/ethereum/vm/VM.java [hunk 2]
```diff
@@ -834,6 +834,41 @@ protected void doCODECOPY() {
         program.step();
     }
 
+    protected void doRETURNDATASIZE() {
+        DataWord dataSize = program.getReturnDataBufferSize();
```
[the remark this answers is not in the document]
Comment (tinchou) [reply rskj-core/src/main/java/org/ethereum/vm/VM.java]: Thanks, great catch!

File: rskj-core/src/main/java/org/ethereum/vm/VM.java (hunk 1 above)
Comment (tinchou) [reply rskj-core/src/main/java/org/ethereum/vm/VM.java]: It's not a new cost, I just extracted a method so I could reuse it from `RETURNDATACOPY `. I'll see if I can add a test for `RETURNDATACOPY `'s gas cost.

File: rskj-core/src/test/java/org/ethereum/vm/ProgramTest.java [hunk 3]
```diff
@@ -85,4 +88,22 @@ public void cantCreateTooLargeContract() {
         Assert.assertNotNull(result.getException());
         Assert.assertTrue(result.getException() instanceof RuntimeException);
```
[the remark this answers is not in the document]
Comment (tinchou) [reply rskj-core/src/test/java/org/ethereum/vm/ProgramTest.java]: Yeah, this is a bigger test, but that would be a good unit test. I'll try to write it.

**Comment by tinchou** [comment]
@ajlopezrsk I think everything should be fine now. Let me know!

File: rskj-core/src/main/java/org/ethereum/vm/program/Program.java [hunk 4]
```diff
@@ -1354,6 +1361,28 @@ public static String stringifyMultiline(byte[] code) {
...
+    public DataWord getReturnDataBufferSize() {
+        return new DataWord(getReturnDataBufferSizeI());
+    }
+
+    private int getReturnDataBufferSizeI() {
+        return returnDataBuffer == null ? 0 : returnDataBuffer.length;
+    }
+
+    public Optional<byte[]> getReturnDataBufferData(DataWord off, DataWord size) {
+        long endPosition = (long) off.intValueSafe() + size.intValueSafe();
```
[the remark this answers is not in the document]
Comment (SergioDemianLerner) [reply rskj-core/src/main/java/org/ethereum/vm/program/Program.java]: The Ethereum Go implementation corresponds to (2):
```
end := new(big.Int).Add(dataOffset, length)
	if end.BitLen() > 64 || uint64(len(evm.interpreter.returnData)) < end.Uint64() {
		return nil, errReturnDataOutOfBounds
	}
```

[the remark this answers is not in the document]
Comment (SergioDemianLerner) [reply rskj-core/src/main/java/org/ethereum/vm/program/Program.java]: THEREFORE, whatever the EIP says, we're implementing the same that is implemented in Ethereum Go code. No problem here.

**Comment by aeidelman** [comment]
@ajlopezrsk please check latest changes

**Pull request merged**